### PR TITLE
Add optional `namespace` for `service` in APIRule

### PR DESCRIPTION
**Pull request by videlov**
**Description**

Added optional `namespace` in `service` for APIRule CRD (on root `spec` or as subelement in `rules`)

Changes in this pull request:

- [x] Update API-Rule CRD
- [x] Adapt target for ory-proxy Ory Rule and upstream for VirtualService
- [x] Unit tests

**Related issue(s)**
https://github.com/kyma-project/kyma/issues/9936
https://github.com/kyma-project/api-gateway/issues/12

### CODE REVIEW

**Review by werdes72 — APPROVED** [review]

**Review by barchw — CHANGES_REQUESTED** [review]
Could you please format with `gofmt -s -w .`?

**Comment by videlov** [comment]
> gofmt -s -w .

done

File: internal/processing/processing_test.go [hunk 1]
```diff
@@ -124,6 +124,44 @@ var _ = Describe("Factory", func() {
 				Expect(len(accessRules)).To(Equal(0))
 			})
 
+			It("should override upstream for specified spec level service namespace", func() {
```
Comment (barchw) [inline internal/processing/processing_test.go:127]:
```suggestion
			It("should override VS destination host for specified spec level service namespace", func() {
```

File: internal/processing/processing_test.go [hunk 2]
```diff
@@ -124,6 +124,44 @@ var _ = Describe("Factory", func() {
...
+
+				desiredState := f.CalculateRequiredState(apiRule)
+				vs := desiredState.virtualService
+				accessRules := desiredState.accessRules
+
+				//verify VS has rule level destination host
+				Expect(len(vs.Spec.Http[0].Route)).To(Equal(1))
+				Expect(vs.Spec.Http[0].Route[0].Destination.Host).To(Equal(overrideServiceName + "." + overrideServiceNamespace + ".svc.cluster.local"))
+
+				//Verify AR has rule level upstream
```
Comment (barchw) [inline internal/processing/processing_test.go:161]:
```suggestion
				//Verify AR does not exist
```

File: internal/processing/processing_test.go [hunk 3]
```diff
@@ -200,6 +280,44 @@ var _ = Describe("Factory", func() {
...
```
Comment (barchw) [inline internal/processing/processing_test.go:317]:
```suggestion
				//Verify AR does not exist
```

File: internal/validation/validate.go [hunk 4]
```diff
@@ -128,7 +127,8 @@ func (v *APIRule) validateService(attributePath string, api *gatewayv1beta1.APIR
 
 	for namespace, services := range v.ServiceBlockList {
 		for _, svc := range services {
-			if svc == *api.Spec.Service.Name && namespace == api.ObjectMeta.Namespace {
+			serviceNamespace := helpers.FindServiceNamespace(api, nil)
```
Comment (barchw) [inline internal/validation/validate.go:130]: `nilpointer dereference` is most likely not possible to occur, but I would handle it just to be  100% sure

File: internal/validation/validate.go [hunk 5]
```diff
@@ -167,7 +168,8 @@ func (v *APIRule) validateRules(attributePath string, checkForService bool, rule
 		if r.Service != nil {
 			for namespace, services := range v.ServiceBlockList {
 				for _, svc := range services {
-					if svc == *r.Service.Name && namespace == rulesNamespace {
+					serviceNamespace := helpers.FindServiceNamespace(api, &r)
```
Comment (barchw) [inline internal/validation/validate.go:171]: > `nilpointer dereference` is most likely not possible to occur, but I would handle it just to be 100% sure

Same here

**Review by barchw — CHANGES_REQUESTED** [review]

**Review by barchw — APPROVED** [review]

**Pull request merged**